File: ObjC/BaseModel.m
//
//  BaseModel.m
//  <>
//
//  Created by Naveen Shan.
//  Copyright © 2015. All rights reserved.
//

#import "BaseModel.h"

#import <objc/runtime.h>

@interface BaseModel ()

@property (nonatomic, strong) NSMutableArray *propertyNames;

@end

@implementation BaseModel

#pragma mark -

- (instancetype)initWithContent:(NSDictionary *)content {
    self = [super init];
    if (self) {
        [self setContents:content];
    }
    return self;
}

+ (NSDictionary *)arrayClassMapping  {
    return nil;
}

+ (NSDictionary *)dictionaryClassMapping  {
    return nil;
}

+ (NSArray *)privateProperties  {
    return nil;
}

+ (NSDictionary *)propertiesMap  {
    return nil;
}

#pragma mark -

- (NSArray *)allPropertyNames   {
    if (self.propertyNames) {
        return self.propertyNames;
    }
    
    unsigned count;
    objc_property_t *properties = class_copyPropertyList([self class], &count);
    
    NSMutableArray *propertyValues = [NSMutableArray array];
    
    unsigned i;
    for (i = 0; i < count; i++) {
        objc_property_t property = properties[i];
        NSString *name = [NSString stringWithUTF8String:property_getName(property)];
        [propertyValues addObject:name];
    }
    NSArray *pvtProperties = [[self class] privateProperties];
    [propertyValues removeObjectsInArray:pvtProperties];
    
    free(properties);
    
    self.propertyNames = propertyValues;
    return self.propertyNames;
}

- (id)convertValueDataTypeToKVOIfRequired:(id)value {
    if (! value) {
        return value;
    }
    if ([value isKindOfClass:[NSArray class]]) {
        value = [[self class] arrayWithModel:value];
    } else if ([value isKindOfClass:[BaseModel class]]) {
        value = [value dictionary];
    }
    return value;
}

- (id)convertValueDataTypeFromKVOIfRequired:(id)value property:(NSString *)property {
    if (! value) {
        return value;
    }

    if ([value isKindOfClass:[NSArray class]]) {
        NSDictionary *arrayMapping = [[self class] arrayClassMapping];
        Class classType = [arrayMapping objectForKey:property];
        if (classType) {
            value = [classType arrayWithArray:value];
        }
    }
    else if ([value isKindOfClass:[NSDictionary class]]) {
        NSDictionary *arrayMapping = [[self class] dictionaryClassMapping];
        Class classType = [arrayMapping objectForKey:property];
        if (classType) {
            value = [[classType alloc] initWithContent:value];
        }
    }
    return value;
}

- (void)setContents:(NSDictionary *)content {
    if (! [content isKindOfClass:[NSDictionary class]]) {
        return;
    }
    NSArray *propertyNames = [self allPropertyNames];
    for (NSString *property in propertyNames) {
        NSString *propertiesMapKey = [[[self class] propertiesMap] objectForKey:property];
        if (!propertiesMapKey) {
            propertiesMapKey = property;
        }
        
        id value = [content valueForKey:propertiesMapKey];
        value = [self convertValueDataTypeFromKVOIfRequired:value property:property];
        if (value && (NSNull *)value != [NSNull null]) {
            [self setValue:value forKey:property];
        }
    }
}

#pragma mark -

- (NSMutableDictionary *)dictionary {
    NSArray *propertyNames = [self allPropertyNames];
    
    NSMutableDictionary *contents = [NSMutableDictionary dictionaryWithCapacity:[propertyNames count]];
    for (NSString *property in propertyNames) {
        id value = [self valueForKey:property];
        
        value = [self convertValueDataTypeToKVOIfRequired:value];
        if (value && (NSNull *)value != [NSNull null]) {
            NSString *propertiesMapKey = [[[self class] propertiesMap] objectForKey:property];
            if (!propertiesMapKey) {
                propertiesMapKey = property;
            }
            [contents setValue:value forKey:propertiesMapKey];
        }
    }
    
    return contents;
}

+ (NSMutableArray *)arrayWithArray:(NSArray *)contents {
    if (! [contents isKindOfClass:[NSArray class]]) {
        return nil;
    }
    NSMutableArray *array = [NSMutableArray array];
    for (NSDictionary *content in contents) {
        id object = [[[self class] alloc] initWithContent:content];
        [array addObject:object];
    }
    return array;
}

+ (NSMutableArray *)arrayWithModel:(NSArray *)contents {
    if (! [contents isKindOfClass:[NSArray class]]) {
        return nil;
    }
    NSMutableArray *array = [NSMutableArray array];
    for (BaseModel *content in contents) {
        if ([content respondsToSelector:@selector(dictionary)]) {
            [array addObject:[content dictionary]];
        } else {
            [array addObject:content];
        }
    }
    return array;
}

+ (instancetype)objectFromDictionary:(NSDictionary *)dictionary {
    return [[self alloc] initWithContent:dictionary];
}

@end

File: Swift/BaseModel.swift
//
//  BaseModel.swift
//  <>
//
//  Created by Naveen Shan.
//  Copyright © 2015. All rights reserved.
//

import Foundation

public class BaseModel: NSObject {
    private var propertyNames: Array<String>?
    
    required public init(content: Dictionary<String, AnyObject>) {
        super.init()
        self.setContents(content)
    }
    
    //MARK: -

    public class func arrayClassMapping() -> Dictionary<String, AnyClass>? {
        return nil
    }
    
    public class func dictionaryClassMapping() -> Dictionary<String, AnyClass>? {
        return nil
    }
    
    public class func privateProperties() -> Array<String>? {
        return nil
    }
    
    public class func propertiesMap() -> Dictionary<String, String>? {
        return nil
    }
    
    //MARK: -

    func allPropertyNames() -> Array<String>? {
        if self.propertyNames != nil {
            return self.propertyNames!
        }
        
        var currentClass: AnyClass = self.dynamicType
        var propertyValues: Array<String> = []

        repeat {
            var count: UInt32 = 0
            let properties = class_copyPropertyList(currentClass, &count);
            for var i = 0; i < Int(count); i++ {
                let property = properties[Int(i)];
                // retrieve the property name by calling property_getName function
                let cname = property_getName(property);
                // covert the c string into a Swift string
                let name = String.fromCString(cname);
                propertyValues.append(name!);
            }
            
            let pvtProperties = currentClass.privateProperties()
            
            propertyValues = propertyValues.filter({ (value: String) -> Bool in
                return !(pvtProperties!.contains(value))
            })
            
            free(properties)
            
            currentClass = currentClass.superclass()!

        } while currentClass != BaseModel.self
        self.propertyNames = propertyValues
        
        return self.propertyNames
    }
    
    
//    func allPropertyNames() -> Array<String>? {
//        if self.propertyNames != nil {
//            return self.propertyNames!
//        }
//        
//        var count: UInt32 = 0
//        let properties = class_copyPropertyList(self.dynamicType, &count);
//        var propertyValues: Array<String> = []
//        for var i = 0; i < Int(count); i++ {
//            let property = properties[Int(i)];
//            // retrieve the property name by calling property_getName function
//            let cname = property_getName(property);
//            // covert the c string into a Swift string
//            let name = String.fromCString(cname);
//            propertyValues.append(name!);
//        }
//        
//        let pvtProperties = self.dynamicType.privateProperties()
//        
//        propertyValues = propertyValues.filter({ (value: String) -> Bool in
//            return !(pvtProperties!.contains(value))
//        })
//        
//        free(properties);
//
//        self.propertyNames = propertyValues
//        return self.propertyNames
//    }
    
    func convertValueDataTypeToKVOIfRequired(var value: AnyObject?) -> AnyObject? {
        if value == nil  {
            return value
        }
        if let aValue = value as? Array<AnyObject> {
            value = self.dynamicType.arrayWithModel(aValue)
        }
        else if let aValue = value as? BaseModel {
            value = aValue.dictionary()
        }
        return value
    }
    
    func convertValueDataTypeFromKVOIfRequired(var value: AnyObject?, property: String) -> AnyObject? {
        if value == nil {
            return value
        }
        if let aValue = value as? Array<AnyObject> {
            let arrayMapping = self.dynamicType.arrayClassMapping()
            let classType: AnyClass = (arrayMapping?[property])!

            if let aClassType = classType as? BaseModel.Type {
                value = aClassType.arrayWithArray(aValue)
            }
        }
        else if let aValue = value as? Dictionary<String, AnyObject> {
            let arrayMapping = self.dynamicType.dictionaryClassMapping()
            let classType: AnyClass = (arrayMapping?[property])!
            
            if let aClassType = classType as? BaseModel.Type {
                value = aClassType.init(content: aValue)
            }
        }
        return value
    }
    
    func setContents(content: Dictionary<String, AnyObject>) {
        let propertyNames = self.allPropertyNames()
        
        for property in propertyNames! {
            var propertiesMapKey = self.dynamicType.propertiesMap()?[property]
            if propertiesMapKey == nil {
                propertiesMapKey = property
            }
            
            let value: AnyObject? = content[propertiesMapKey!]
            
            if (value != nil && !(value is NSNull)) {

                self.setValue(value, forKey: property)
            }
        }
    }
    
    //MARK: -
    
    public func dictionary() -> Dictionary<String, AnyObject> {
        let propertyNames = self.allPropertyNames()
        var contents = Dictionary<String, AnyObject>(minimumCapacity: propertyNames!.count)
        for property in propertyNames! {
            var value = self.valueForKey(property)
            value =  self.convertValueDataTypeToKVOIfRequired(value)
            
            if value != nil && !(value is NSNull) {
                var propertiesMapKey = self.dynamicType.propertiesMap()?[property]
                if propertiesMapKey == nil {
                    propertiesMapKey = property
                }
                contents[propertiesMapKey!] = value
            }
        }
        return contents
    }

    public class func arrayWithArray(contents: Array<AnyObject>) -> Array<AnyObject>? {
        var array: Array<AnyObject> = []
        for content in contents {
            let dict = content as? Dictionary<String, AnyObject>
            let object = BaseModel(content: dict!)
            array.append(object)
        }
        return array
    }
    
    class func arrayWithModel(contents: Array<AnyObject>) -> Array<AnyObject>? {
        var array: Array<AnyObject> = []
        for content in contents {
            let model = content as? BaseModel
            if model?.dictionary != nil {
                array.append(model!.dictionary())
            }
            else {
                array.append(content)
            }
        }
        return array
    }
    
    public class func objectFromDictionary(dictionary: Dictionary<String, AnyObject>) -> Self {
        return self.init(content: dictionary)
    }
    

    
}







